### ISSUE TITLE
weird element placement behavior

### ISSUE THREAD

**Issue by IlhamWahabiGX** [stereobooster/react-snap#214]
after using react snap (there is no problem before), instead got this normal result:
![before](https://puu.sh/AZcWb/e9479a5521.png)

with this normal structure
![before](https://puu.sh/AZcU8/1d41287765.png)

---
i got this weird result
![after](https://puu.sh/AZd1I/f852e77559.png)

with this weid structure
![after](https://puu.sh/AZd1i/385fd392b4.png)

---

As you can see, the 'contact__item' is supposed inside 'contact' but instead I got it's inside 'jss20'. Its also got applied in 'home__desc'

I'm using react-reveal for animate the component and at first I think its the cause, but the result still same even when I'm remove the animation.

The fun part is, when I click the link to move to another route and back to home its become normal again. So i think the problem is with prerendered html

Already tested in latest chrome and mozilla with hard reload.

If you are curious with how i use css, and its same with another component
![code](https://puu.sh/AZdfN/3cfc1e6996.png)

**Comment by stereobooster** [comment]
Hi @IlhamWahabiGX. It is hard to tell what exactly cause this behaviour I haven't seen anything like that before, can you try to construct minimal example where it is possible to reproduce bug. It is obvious that there is unclosed html tag somewhere, but I'm not sure why it happens.

**Comment by bfricka** [comment]
This is also happening for me. I'm trying to narrow down where the unclosed tag could be.

**Comment by stereobooster** [comment]
@bfricka it can happen that hydration issue and unclosed tag are not connected. Can you post what hydration exception saying?

> Hydration fails when it hits Foo because as a child, the type property of this fiber is undefined, meaning it's a component.

Not sure what you mean. Hydration works in [this example](https://github.com/stereobooster/an-almost-static-stack/blob/react-snap/src/App.js#L83), which is pretty similar to what you post. We can create example with the same versions of libs as yours (React etc).

**Comment by bfricka** [comment]
That's a poor example b/c it's using custom scripts in a non-ejected `create-react-app` project. As mentioned, you need to get to a state that uses `react-dom.development.js` so  you can see thrown hydration errors.

Currently working on some deeper debugging. I used create-react-app + the steps I outlined above to prove that there are lots of hydration warnings for trivial examples, but none of them have hit what I believe to be the more critical failure points: `didNotHydrateInstance`, `didNotFindHydratableInstance`, and similar. It's easy to replicate for invalid text nodes or classnames being out of order, but my hunch is that these aren't problematic for hydration.

The crucial thing I think I'm seeing is in `tryToClaimNextHydratableInstance` and `tryHydrate` fails. For example in one case that I'm having issues with, the fiber thinks it should be a `ul` when it clearly should be `div`. This fiber has the correct props / className, etc, but it's somehow a completely different tag. There's no conditional logic in this component that might change this dynamically, either.

But basically, when `tryHydrate` fails it will either insert a node by trying to hydrate the next sibling, or, it that doesn't work, it will delete the first one it tried from the parent. Here's the exact code, which seems a bit fluffy:

```js
  if (!tryHydrate(fiber, nextInstance)) {
    // If we can't hydrate this instance let's try the next one.
    // We use this as a heuristic. It's based on intuition and not data so it
    // might be flawed or unnecessary.
    nextInstance = getNextHydratableSibling(firstAttemptedInstance);
    if (!nextInstance || !tryHydrate(fiber, nextInstance)) {
      // Nothing to hydrate. Make it an insertion.
      insertNonHydratedInstance(hydrationParentFiber, fiber);
      isHydrating = false;
      hydrationParentFiber = fiber;
      return;
    }
    // We matched the next one, we'll now assume that the first one was
    // superfluous and we'll delete it. Since we can't eagerly delete it
    // we'll have to schedule a deletion. To do that, this node needs a dummy
    // fiber associated with it.
    deleteHydratableInstance(hydrationParentFiber, firstAttemptedInstance);
  }
  hydrationParentFiber = fiber;
  nextHydratableInstance = getFirstHydratableChild(nextInstance);
}
```

The real question is why this discrepancy. I have a few hypotheses to try and I'll report back if I find anything useful.

**Comment by stereobooster** [comment]
You want to say if I will eject and replace production with development I will start to see hydration errors - they exist, but I don't see it because of production build instead of development. I will try.

**Comment by bfricka** [comment]
I haven't fixed it yet, but I believe I found the source of my particular issue. I use react-i18next for translations, and this will check to see if translations are loaded before rendering its children. I followed the render loop to the problem area and found that when it gets to this part of the tree and calls the render function on `I18n`, it returns null, thus creating the discrepancy.

I found that all the hydration warnings I mentioned previously are still going to be there, but hydrate itself doesn't fail unless `tryHydrate` fails. Additionally, the more important issue here is having a different DOM output. In my case, it's an easy fix. I can either turn off waiting (`{react: {wait: false}}`), or preload the initial translations.

**Comment by stereobooster** [comment]
> or preload the initial translations.

You can save translations with `snapSaveState`.

**Comment by stereobooster** [comment]
@bfricka you are totally right about `sortClassName` it should be false otherwise hydration doesn't work. This is fixed in v.1.17.0. But I do not see any other issues except this one. See [my test repo](https://github.com/stereobooster/an-almost-static-stack/tree/hydration-error)

**Comment by bfricka** [comment]
Yeah, I think we're on the same page. It wasn't until I stepped through everything that I saw that there are different issues associated w/ hydration. Some of these issues are minor and only throw warnings, while hydration continues (className / attribute order, text content changes, etc) successfully. Other issues, like those caused by dynamic content mismatches (i18n, and I presume redux state issues), will cause hydration to bail.

Thanks for engaging in the conversation here. I'm not sure if this addresses OP's question or not, but I would presume it has something to do w/ dynamic state missing on initial reconciliation.

Thanks for the `snapSaveState` suggestion. I already use this for pre-populating the redux state, but I'm not sure about using it for translations. `react-i18next` provider allows an the initial translations to be passed statically at run time, but it may be useful to use `snapSaveState` if I ever want to render different versions for each language. If I did do this, I'd have to figure out how to get puppeteer to present `navigator.language` / `navigator.languages`. 

Thanks again.
